core/src/main/java/brooklyn/entity/rebind/RebindIteration.java

```diff
@@ -899,48 +914,64 @@ protected void setCatalogItemId(BrooklynObject item, String catalogItemId) {
-                }
-
-                if (BrooklynFeatureEnablement.isEnabled(BrooklynFeatureEnablement.FEATURE_INFER_CATALOG_ITEM_ON_REBIND)) {
-                    //Try loading from whichever catalog bundle succeeds.
-                    BrooklynCatalog catalog = managementContext.getCatalog();
-                    for (CatalogItem<?, ?> item : catalog.getCatalogItems()) {
-                        BrooklynClassLoadingContext catalogLoader = CatalogUtils.newClassLoadingContext(managementContext, item);
-                        Maybe<Class<?>> catalogClass = catalogLoader.tryLoadClass(jType);
-                        if (catalogClass.isPresent()) {
-                            return (Class<? extends T>) catalogClass.get();
+                CatalogItem<?, ?> catalogItem = rebindContext.lookup().lookupCatalogItem(catalogItemId);
+                if (catalogItem == null) {
+                    if (BrooklynFeatureEnablement.isEnabled(BrooklynFeatureEnablement.FEATURE_AUTO_FIX_CATALOG_REF_ON_REBIND)) {
+                        // See https://issues.apache.org/jira/browse/BROOKLYN-149
+                        // This is a dangling reference to the catalog item (which will have been logged by lookupCatalogItem).
+                        // Try loading as any version.
+                        if (CatalogUtils.looksLikeVersionedId(catalogItemId)) {
+                            String symbolicName = CatalogUtils.getIdFromVersionedId(catalogItemId);
+                            catalogItem = rebindContext.lookup().lookupCatalogItem(symbolicName);
+                            
+                            if (catalogItem != null) {
+                                LOG.warn("Unable to load catalog item "+catalogItemId+" for "+contextSuchAsId
+                                        +" ("+bType.getSimpleName()+"); will auto-upgrade to "+catalogItem.getCatalogItemId());
+                                catalogItemId = catalogItem.getCatalogItemId();
+                            }
                         }
                     }
-                    throw new IllegalStateException("No catalogItemId specified and can't load class from either classpath of catalog items");
+                }
+                if (catalogItem != null) {
+                    BrooklynClassLoadingContext loader = CatalogUtils.newClassLoadingContext(managementContext, catalogItem);
+                    return new LoadedClass<T>(loader.loadClass(jType, bType), catalogItemId);
                 } else {
-                    throw new IllegalStateException("No catalogItemId specified and can't load class from classpath");
+                    LOG.warn("Unable to load catalog item "+catalogItemId+" for "+contextSuchAsId
+                            +" ("+bType.getSimpleName()+"); will try default class loader");
                 }
-
             }
-        }
+            
+            try {
+                return new LoadedClass<T>((Class<T>)reflections.loadClass(jType), catalogItemId);
```

Comment:
Reason I didn't return null was that it would override the `catalogItemId` to be persisted, so it would then say null. I didn't want to lose the catalogItemId in the situation where it wasn't available. Instead, we'd leave that itemId in and fall back to default class loader.

I wondered about guarding it behind the feature enablement. However, neither of the (renamed) feature options seem quite right:
* `FEATURE_BACKWARDS_COMPATIBILITY_INFER_CATALOG_ITEM_ON_REBIND` is about backwards compatibility, where the `catalogItemId` was missing from old things.
* `FEATURE_AUTO_FIX_CATALOG_REF_ON_REBIND` is for updating the ref (I think that deleting the ref seems extreme there).

We should revisit this in follow-up pull requests. I'd like us to start up such that errors are shown, and can be fixed, rather than aborting entirely or just logging and ignoring.